service/src/event/order_events.rs

```diff
@@ -47,25 +47,31 @@ impl OrderEvents {
                 (e.cid(), e.deliverable())
             }));
         let mut deliverable = Vec::with_capacity(candidate_events.len());
-        candidate_events.retain(|(e, h)| {
+        let mut remaining_candidates = Vec::with_capacity(candidate_events.len());
+
+        while !candidate_events.is_empty() {
+            let (e, h) = candidate_events.pop().unwrap();
             if e.deliverable() {
-                deliverable.push((e.clone(), h.clone()));
-                false
+                deliverable.push((e, h))
             } else {
-                true
+                remaining_candidates.push((e, h))
             }
-        });
-        if candidate_events.is_empty() {
+        }
+        // Need to reverse the vectors because we popped events off of candidate_events from back to front
+        deliverable.reverse();
```

Comment:
this feels a bit hacky and probably not the most efficient.  The other idea I had was to replace the `pop().unwrap()` with `remove(0)`, but I was worried that removing from the front of the array repeatedly would be even less efficient than reversing at the end.

It really feels like there should be a better way to move elements out of one vector into two others in an efficient manor...

Reply:
Adding @nathanielc to see if you have any ideas for how to do this more efficiently

Reply:
VecDeque and push front.

Reply:
What does this mean?  I have a vector, I want to move the elements from that vector out into two other vectors, deciding which vector to move into based on a property of the values in the input vector.  How can a VecDeque help with that?

Reply:
Instead of using a vector to store events, use a VecDeque. You push front instead of back. You then return an object with a VecDeque.

Reply:
Hmm but at the end I need to be returning a Vec, so I'd need to move everything out of the VecDeque into a Vec at the end anyway, is that actually going to be any more efficient?

Reply:
Make sure I understand. We have three vectors:

* candidate_events
* deliverable
* remaining_candidates

We want to _move_ the values from candidate_events into one of the other two vectors and we want to do it in order? When we are done candidate_events is empty?

In this case a simple for loop should suffice.

```rust
for event in candidate_events {
   if ... {deliverable.push(event)} else {remaining_candidates.push(event) }
}
```
Note since we are not using a reference to candidate_events in the for loop the values are moved out of the vector into the loop.

Aside there is a one liner for this in nightly rust https://doc.rust-lang.org/std/vec/struct.Vec.html#method.extract_if

Reply:
This looks like a nice cleanup, but I'd like to do that in a separate PR to keep this one more focused.

Reply:
Made a ticket to revisit this: https://linear.app/3boxlabs/issue/AES-309/cleanup-ordereventstry-new